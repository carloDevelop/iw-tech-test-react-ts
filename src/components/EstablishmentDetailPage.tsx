import { useState, useEffect } from "react";
import { useParams, useNavigate } from "react-router-dom";
import styles from "../styles/EstablishmentDetailPage.module.css";
import {
  getEstablishmentDetails,
  EstablishmentDetailType,
} from "../api/ratingsAPI";
import LoadingSpinner from "./LoadingSpinner";
import GenericButton from "./GenericButton";

export const EstablishmentDetailPage = () => {
  const { id } = useParams<{ id: string }>();
  const navigate = useNavigate();
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<{
    message: string;
    [key: string]: string;
  }>();
  const [details, setDetails] = useState<EstablishmentDetailType>();

  useEffect(() => {
    getEstablishmentDetails(id)
      .then(
        (result) => {
          setDetails(result);
        },
        (error) => {
          setError(error);
        }
      )
      .finally(() => {
        setLoading(false);
      });
  }, []);

  const navigateBack = () => {
    navigate("/");
  };
  const formatedDate = (date: string | undefined) => {
    return date ? new Date(date).toLocaleDateString() : "N/A";
  };

  const displayRating = (values: {} | undefined) => {
    const entries = values ? Object.entries(values) : [];
    return (
      <ul>
        {entries.map((entry) => (
          <li key={entry[0]}>
            <p> {entry[0]}:  {entry[1]}</p>
          </li>
        ))}
      </ul>
    );
  };

  return (
    <div className={styles.detailTable}>
      {loading ? (
        <LoadingSpinner message="Loading..." />
      ) : (
        <div data-testid="detailPage" className={styles.details}>
          <h1> {details?.BusinessName} </h1>
          <p data-testid="fullAddress">
            <b>Address:</b>{` ${details?.AddressLine1} ${details?.AddressLine2}${details?.AddressLine3.toString() === "" ? "," : " "+details?.AddressLine3+","} ${details?.PostCode} ${details?.AddressLine4}`}
          </p>
          <p data-testid="dateOfInspection">
            <b>Date of inspection:</b> {formatedDate(details?.RatingDate)}
          </p>
          <div>
          <p data-testid="rating"><b>Rating:</b>{" "}{details?.RatingValue}</p>
          </div>
          {displayRating(details?.scores)}

          <GenericButton text="Go Back" onClick={navigateBack} />
        </div>
      )}
    </div>
  );
};
