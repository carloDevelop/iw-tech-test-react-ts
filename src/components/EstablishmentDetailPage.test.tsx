import { render, screen, waitFor, act } from "@testing-library/react";
import { EstablishmentDetailPage } from "./EstablishmentDetailPage";
import { MemoryRouter } from "react-router-dom";
import { enableFetchMocks } from "jest-fetch-mock";
import fetch from "jest-fetch-mock";

describe("EstablishmentDetailPage", () => {
  beforeEach(() => {
    enableFetchMocks();
    fetch.resetMocks();
  });

  it("renders EstablishmentDetailPage and checks loading indicator", () => {
    act(() => {
      render(
        <MemoryRouter initialEntries={["/detail/123"]}>
          <EstablishmentDetailPage />
        </MemoryRouter>
      );
    })
    
    const spinner = screen.getByRole("status");
    expect(spinner).toBeInTheDocument();
  });

  it("renders EstablishmentDetailPage with details", async () => {
    const expected = {
      AddressLine1: "Camden Street",
      AddressLine2: "192-198 Camden",
      AddressLine3: "",
      AddressLine4: "London",
      BusinessName: "Test Establisment",
      PostCode: "123456",
      RatingDate: "2023-10-12T00:00:00",
      RatingValue: "2",
      scores: {
        Hygiene: 15,
        Structural: 10,
        ConfidenceInManagement: 10,
      },
    };
    fetch.mockResponseOnce(JSON.stringify(expected));
    
    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/detail/123"]}>
          <EstablishmentDetailPage />
        </MemoryRouter>
      );
      await waitFor(() => expect(screen.getByTestId("detailPage")).toBeInTheDocument());
    })
       
    expect(screen.getByTestId("detailPage")).toMatchSnapshot();
    expect(screen.getByTestId("fullAddress").textContent).toEqual(
      "Address: Camden Street 192-198 Camden, 123456 London"
    );
    expect(screen.getByTestId("dateOfInspection").textContent).toEqual(
      "Date of inspection: 10/12/2023"
    );
    expect(screen.getByTestId("rating").textContent).toEqual("Rating: 2");
  });
});
