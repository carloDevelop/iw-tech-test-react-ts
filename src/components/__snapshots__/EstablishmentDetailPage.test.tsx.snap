// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstablishmentDetailPage renders EstablishmentDetailPage with details 1`] = `
<div
  class="details"
  data-testid="detailPage"
>
  <h1>
     
    Test Establisment
     
  </h1>
  <p
    data-testid="fullAddress"
  >
    <b>
      Address:
    </b>
     Camden Street 192-198 Camden, 123456 London
  </p>
  <p
    data-testid="dateOfInspection"
  >
    <b>
      Date of inspection:
    </b>
     
    10/12/2023
  </p>
  <div>
    <p
      data-testid="rating"
    >
      <b>
        Rating:
      </b>
       
      2
    </p>
  </div>
  <ul>
    <li>
      <p>
         
        Hygiene
        :  
        15
      </p>
    </li>
    <li>
      <p>
         
        Structural
        :  
        10
      </p>
    </li>
    <li>
      <p>
         
        ConfidenceInManagement
        :  
        10
      </p>
    </li>
  </ul>
  <button
    class="button"
  >
    Go Back
  </button>
</div>
`;
